tools/clang/lib/AST/ASTContextHLSL.cpp

```diff
@@ -478,48 +512,18 @@ void hlsl::AddHLSLVectorTemplate(ASTContext& context, ClassTemplateDecl** vector
     context.DeclarationNames.getCXXOperatorName(OO_Subscript), ForConstFalse);
   AddHLSLVectorSubscriptAttr(functionDecl, context);
 
-  templateRecordDecl->completeDefinition();
-
-  classTemplateDecl->setImplicit(true);
-  templateRecordDecl->setImplicit(true);
-
-  // Both declarations need to be present for correct handling.
-  currentDeclContext->addDecl(classTemplateDecl);
-  currentDeclContext->addDecl(templateRecordDecl);
-
-#ifdef DBG
-  // Verify that we can read the field member from the template record.
-  DeclContext::lookup_result lookupResult = templateRecordDecl->lookup(
-    DeclarationName(&context.Idents.get(StringRef("h"))));
-  DXASSERT(!lookupResult.empty(), "otherwise vector handle cannot be looked up");
-#endif
-
+  typeDeclBuilder.completeDefinition();
   *vectorTemplateDecl = classTemplateDecl;
 }
 
 /// <summary>
 /// Adds a new record type in the specified context with the given name. The record type will have a handle field.
 /// </summary>
-void hlsl::AddRecordTypeWithHandle(ASTContext& context, _Outptr_ CXXRecordDecl** typeDecl, _In_z_ const char* typeName)
-{
-  DXASSERT_NOMSG(typeDecl != nullptr);
-  DXASSERT_NOMSG(typeName != nullptr);
-  
-  *typeDecl = nullptr;
-
-  DeclContext* currentDeclContext = context.getTranslationUnitDecl();
-  IdentifierInfo& newTypeId = context.Idents.get(StringRef(typeName), tok::TokenKind::identifier);
-  CXXRecordDecl* newDecl = CXXRecordDecl::Create(
-    context, TagDecl::TagKind::TTK_Struct, currentDeclContext, NoLoc, NoLoc, &newTypeId, nullptr);
-  newDecl->setLexicalDeclContext(currentDeclContext);
-  newDecl->setFreeStanding();
```

Comment:
This `setFreeStanding` is the only thing my code does differently, in that it won't call it. This is the only function in this file that used this method, and according to the documentation it's used to mark forward declarations like `struct Foo;`, so I think it was bogus.